doc/getdns_dict.3.in
.\" The "BSD-New" License
.\" 
.\" Copyright (c) 2013, NLnet Labs, Verisign, Inc.
.\" All rights reserved.
.\" 
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions are met:
.\" * Redistributions of source code must retain the above copyright
.\"   notice, this list of conditions and the following disclaimer.
.\" * Redistributions in binary form must reproduce the above copyright
.\"   notice, this list of conditions and the following disclaimer in the
.\"   documentation and/or other materials provided with the distribution.
.\" * Neither the names of the copyright holders nor the
.\"   names of its contributors may be used to endorse or promote products
.\"   derived from this software without specific prior written permission.
.\" 
.\" THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
.\" WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
.\" DISCLAIMED. IN NO EVENT SHALL Verisign, Inc. BE LIABLE FOR ANY
.\" DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
.\" (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
.\" LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
.\" ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
.\" (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
.\" SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
.\" 

.TH getdns_dict 3 "@date@" "getdns @version@" getdns
.ad l
.SH NAME
.B getdns_dict, 
.B getdns_dict_create, 
.B getdns_dict_create_with_extended_memory_functions, 
.B getdns_dict_create_with_memory_functions, 
.B getdns_dict_destroy 
-- getdns dict create and destroy routines
.ad n

.SH LIBRARY
DNS Resolver library (libgetdns, -lgetdns)

.SH SYNOPSIS
#include <getdns.h>

getdns_dict * 
.br
.B getdns_dict_create
()

getdns_dict *
.br
.B getdns_dict_create_with_extended_memory_functions
.RS 3
(void *userarg,
.br
void *(*malloc)(void *userarg, size_t sz),
.br
void *(*realloc)(void *userarg, void *buf, size_t sz),
.br
void (*free)(void *userarg, void *buf)
.RE 3

getdns_dict *
.br
.B getdns_dict_create_with_memory_functions
.RS 3
(void *(*malloc)(size_t sz),
.br
void *(*realloc)(void *buf, size_t sz),
.br
void (*free)(void *buf)
.RE 3

void 
.br
.B getdns_dict_destroy
(getdns_dict *this_dict)

.SH DESCRIPTION

.LP
The getdns_dict type is used to manage name/value pairs in which the names are strings and the data types of the values are heterogeneous and include
.RS 3
.br
getdns_bindata
.br
getdns_dict
.br
getdns_list
.br
uint32_t
.RE

.LP
The destroy function performs a "deep" destroy, freeing storage for all of the members
of the dictionary before destroying the dictionary.  There are a number of helper
functions that provide access to the dictionary object, see their respective man pages.

.LP
.I userarg
pass this argument to the user specified memory management functions for operations on lists created using extended memory functions 
.LP
.I this_dict
the dictionary to destroy

.SH "RETURN VALUES"

Upon successful completion the getdns_dict_create function returns a valid
(empty) dictionary structure that should be freed via a call to
getdns_dict_destroy.  If a parameter in invalid or in the event of some error
getdns_dict_create returns NULL.

.SH EXAMPLES

TBD

.SH SEE ALSO
.BR libgetdns (3),
.BR getdns_dict_get (3),
.BR getdns_dict_set (3),
.BR getdns_list (3),
.BR getdns_pretty_print_dict (3).



ChangeLog
* 2015-01-16: Version 0.1.6
  * Fix: linking against libev on FreeBSD
  * Fix: Let configure report problem on FreeBSD when configuring with
    libevent and libunbound <= 1.4.22 is not compiled with libevent.
  * Fix: Build on Mac OS-X
  * Fix: Lintian errors in manpages
  * Better libcheck detection
  * Better portability with UNIX systems

* 2014-10-31: Version 0.1.5
  * Unit tests for transport settings
  * Fix: adhere to set maximum UDP payload size
  * API change: when no maximum UDP payload size is set, outgoing
    values will adhere to the suggestions in RFC 6891 and may follow
    a scheme that uses multiple values to maximize receptivity.
  * Stub mode use 1232 maximum UDP payload size when connecting to an
    IPv6 upstreams and 1432 with an IPv4 upstream.
  * Evaluate namespaces (or not) on a per query basis
  * GETDNS_NAMESPACE_LOCALNAMES namespace now gives just_address_answers
    only and does not mimic a DNS packet answer anymore
  * The add_opt_parameters extension
  * IPv6 scope_id support with link-local addresses.  Both with parsing
    /etc/resolv.conf and by providing them explicitly via
    getdns_context_set_upstream_recursive_servers
  * Query for A and AAAA simultaneously with return_both_v4_and_v6
  * GETDNS_TRANSPORT_TCP_ONLY_KEEP_CONNECTIONS_OPEN DNS transport
  * Fix: Answers without RRs in query secion (i.e. REFUSED)
  * Fix: Return empty response dict on timeout in async mode too
  * Move spec examples to spec subdirectory
  * Fix issue#76: Setting UDP Payload size below 512 should not error
  * Fix: Include OPT RR in response dict always (even without options)
  * TCP Fast open support (linux only).
    Enable with the --enable-tcp-fastopen configure option
  * Bump library version because of binary API change

* 2014-09-03: Version 0.1.4
  * Synchronous resolves now respect timeout setting,
  * On timeout *_sync functions now return GETDNS_RETURN_GOOD and a
    response dict with "status" GETDNS_RESPSTATUS_ALL_TIMEOUT>
  * Fix issue#50: getdns_dict_remove_name returns GETDNS_RETURN_GOOD on
    success.
  * Fix Issue#54: set_ub_dns_transport() not working
  * Fix Issue#49: Typo in documentation (thanks Stephane Bortzmeyer)
  * getdns_context_set_limit_outstanding_queries(),
    getdns_context_set_dnssec_allowed_skew() and
    getdns_context_set_edns_maximum_udp_payload_size() now working
  * <rr>_unknown rdata field for unknown or unsupported RR types
  * Temporarily disable timeout unit test 3 because of unpredictable results
  * Spec updated to version 0.507
  * Renamed "resolver_type" to "resolution_type" in dict returned from
    getdns_context_get_api_information()
  * Added GETDNS_RESPSTATUS_ALL_BOGUS_ANSWERS return code for with the
    dnssec_return_only_secure extension
  * Added support for CDS and CDNSKEY RR types, but needs ldns > 1.6.17 to
    be able to parse the wire format (not released yet at time of writing)
  * Added OPENPGPKEY RR type, but no rdata fields implementation yet
  * Updated spec to version 0.508 (September 2014)
  * Also chase NSEC and NSEC3 RRSIGs with dnssec_return_validation_chain

* 2014-06-25: Version 0.1.3
  * libtool chage, remove -release, added -version-info
  * Update specification to the June 2014 version (0.501)

* 2014-06-02: Version 0.1.2
  * Fixed rdata fields for MX
  * Expose only public API symbols
  * Updated manpages
  * specify_class extension
  * Build from separate build directory
  * Anticipate libunbound not returning the answer packet
  * Pretty print bindata's representing IP addresses
  * Anticipate absense of implicit DSO linking
  * Mention getdns specific options to configure in INSTALL
    Thanks Paul Hoffman
  * Mac OSX package built instructions for generic user in README.md
    Thanks Joel Purra
  * Fixed build problems on RHEL/CentOS due using libevent 1.x


* 2014-03-24 : Version 0.1.1
  * default to NOT build extensions (libev, libuv, libevent), handle
    --with/--without options to configure for them
  * Fixed some build/make nits
  * respect configure --docdir=X
  * Documentation/man page updates
  * Fix install and cpp guards in getdns_extra.h
  * Add method to switch between threads and fork mode for unbound
  * Fixes for libuv integration (saghul)
  * Fixes for calling getdns_destroy_context within a callback
  * Fixed signal related defines/decls


* 2014-02-25 : Version 0.1.0
  * Initial public release of the getdns API


src/common/getdns_dict.c
/**
 *
 * getdns list management functions, note that the internal storage is 
 * accomplished via the libc binary search tree implementation so your
 * pointer foo needs to be keen to digest some of the internal semantics
 * 
 * Interfaces originally taken from the getdns API description pseudo implementation.
 *
 */
/* The MIT License (MIT)
 * Copyright (c) 2013 Verisign, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */

#include <stdio.h>
#include <search.h>
#include <string.h>
#include <getdns_core_only.h>

/* TODO: change this to make the walk safe for reentrant/multi-thread calls */
struct getdns_list *walkresultlist;
struct getdns_dict *walkresultdict;
char *walkresultchar;
int  walkresultcharlen;

/*---------------------------------------- getdns_dict_cmp */
/**
 * private function used by the t*() functions for managing binary trees
 * behaves similar to strcmp() 
 * @param itemp1 pointer to pointer to getdns_dict_item to compare
 * @param itemp2 pointer to pointer to getdns_dict_item to compare
 * @return results of lexicographic comparison between item1->key and item2->key
 */
int
getdns_dict_cmp(const void *item1, const void *item2)
{ 
    int retval = 0;

    if(item1 == NULL)
    {
        if(item2 == NULL)
            retval = 0;
        else
            retval = -1;
    }
    else if(item2 == NULL)
        retval = 1;
    else
    {
        retval = strcmp(((struct getdns_dict_item *) item1)->key
         , ((struct getdns_dict_item *) item2)->key);
    }

    return retval;
} /* getdns_dict_comp */

/*---------------------------------------- getdns_dict_find */
/**
 * private function used to locate a key in a dictionary
 * @param dict dicitonary to search
 * @param key key to search for
 * @param addifnotfnd if TRUE then an item will be added if the key is not found
 * @return pointer to dictionary item, caller must not free storage associated with item
 * @return NULL if additnotfnd == FALSE and key is not in dictionary
 */
struct getdns_dict_item *
getdns_dict_find(struct getdns_dict *dict, char *key, bool addifnotfnd)
{ 
    struct getdns_dict_item keyitem;
    struct getdns_dict_item **item;
    struct getdns_dict_item *newitem;
    struct getdns_dict_item *ret = NULL;

    if(dict != NULL && key != NULL)
    {
        /* we try to find it first, if we do then clear the existing data */
        keyitem.key    = key;
        keyitem.dtype  = t_invalid;
        keyitem.data.n = 0;
        item = tfind(&keyitem, &(dict->rootp), getdns_dict_cmp);
        if(addifnotfnd == true && (item == NULL || *item == NULL))
        {
            /* tsearch will add a node automatically for us */
            newitem = (struct getdns_dict_item *) malloc(sizeof(struct getdns_dict_item));
            newitem->key    = strdup(key);
            newitem->dtype  = t_invalid;
            newitem->data.n = 0;
            item = tsearch(newitem, &(dict->rootp), getdns_dict_cmp);
        }
        if(item != NULL)
            ret = *item;
    }

    return ret;
} /* getdns_dict_find */

/*---------------------------------------- getdns_dict_visit */
/**
 * private function called by the tree walk function invoked by getdns_dict_get_names
 * it is called as each node is visited.  twalk() calls 3x for each node and passes order
 * to tell us whether it is a pre/in/post order
 */
void
getdns_dict_visit(const void *node, VISIT order, int level)
{
    struct getdns_dict_item *item;
    size_t index;

    item = *(struct getdns_dict_item **) node;
    /* postorder is mis-named - it results in in-order traversal */
    if(order == postorder || order == leaf)
    {
        if(getdns_list_add_item(walkresultlist, &index) == GETDNS_RETURN_GOOD)
        {
            switch(item->dtype)
            {
                case t_bindata:
                    getdns_list_set_bindata(walkresultlist, index, item->data.bindata);
                    break;

                case t_dict:
                    getdns_list_set_dict(walkresultlist, index, item->data.dict);
                    break;

                case t_int:
                    getdns_list_set_int(walkresultlist, index, item->data.n);
                    break;

                case t_list:
                    getdns_list_set_list(walkresultlist, index, item->data.list);
                    break;

                case t_invalid:
                default:
                    // TODO: this is a fault of some kind, for now ignore it
                    break;
            }
        }
    }

    return;
} /* getdns_dict_visit */

/*---------------------------------------- getdns_dict_get_names
 TODO: this needs to be made thread safe by creating a thread specific list
 the binary search tree implementation in the 
*/
getdns_return_t
getdns_dict_get_names(struct getdns_dict *dict, struct getdns_list **answer)
{
    getdns_return_t retval = GETDNS_RETURN_NO_SUCH_DICT_NAME;

    if(dict != NULL && answer != NULL)
    {
        *answer = getdns_list_create();
        walkresultlist = *answer;

        twalk(dict->rootp, getdns_dict_visit);

        retval = GETDNS_RETURN_GOOD;
    }

    return retval;
} /* getdns_dict_get_names */

/*---------------------------------------- getdns_dict_get_data_type */
getdns_return_t
getdns_dict_get_data_type(struct getdns_dict *dict, char *name, getdns_data_type *answer)
{ 
    struct getdns_dict_item *item;
    getdns_return_t retval = GETDNS_RETURN_NO_SUCH_DICT_NAME;

    if(dict != NULL && name != NULL && answer != NULL)
    {
        item = getdns_dict_find(dict, name, false);
        if(item != NULL)
        {
            *answer = item->dtype;
            retval = GETDNS_RETURN_GOOD;
        }
    }

    return retval;
} /* getdns_dict_get_data_type */

/*---------------------------------------- getdns_dict_get_dict */
getdns_return_t
getdns_dict_get_dict(struct getdns_dict *dict, char *name, struct getdns_dict **answer)
{ 
    struct getdns_dict_item *item;
    getdns_return_t retval = GETDNS_RETURN_NO_SUCH_DICT_NAME;

    if(dict != NULL && name != NULL && answer != NULL)
    {
        item = getdns_dict_find(dict, name, false);
        if(item != NULL)
        {
            if(item->dtype != t_dict)
                retval = GETDNS_RETURN_WRONG_TYPE_REQUESTED;
            else
            {
                *answer = item->data.dict;
                retval = GETDNS_RETURN_GOOD;
            }
        }
    }

    return retval;
} /* getdns_dict_get_dict */

/*---------------------------------------- getdns_dict_get_list */
getdns_return_t
getdns_dict_get_list(struct getdns_dict *dict, char *name, struct getdns_list **answer)
{ 
    struct getdns_dict_item *item;
    getdns_return_t retval = GETDNS_RETURN_NO_SUCH_DICT_NAME;

    if(dict != NULL && name != NULL && answer != NULL)
    {
        item = getdns_dict_find(dict, name, false);
        if(item != NULL)
        {
            if(item->dtype != t_list)
                retval = GETDNS_RETURN_WRONG_TYPE_REQUESTED;
            else
            {
                *answer = item->data.list;
                retval = GETDNS_RETURN_GOOD;
            }
        }
    }

    return retval;
} /* getdns_dict_get_list */

/*---------------------------------------- getdns_dict_get_bindata */
getdns_return_t
getdns_dict_get_bindata(struct getdns_dict *dict, char *name, struct getdns_bindata **answer)
{ 
    struct getdns_dict_item *item;
    getdns_return_t retval = GETDNS_RETURN_NO_SUCH_DICT_NAME;

    if(dict != NULL && name != NULL && answer != NULL)
    {
        item = getdns_dict_find(dict, name, false);
        if(item != NULL)
        {
            if(item->dtype != t_bindata)
                retval = GETDNS_RETURN_WRONG_TYPE_REQUESTED;
            else
            {
                *answer = item->data.bindata;
                retval = GETDNS_RETURN_GOOD;
            }
        }
    }

    return retval;
} /* getdns_dict_get_bindata */

/*---------------------------------------- getdns_dict_get_int */
getdns_return_t
getdns_dict_get_int(struct getdns_dict *dict, char *name, uint32_t *answer)
{ 
    struct getdns_dict_item *item;
    getdns_return_t retval = GETDNS_RETURN_NO_SUCH_DICT_NAME;

    if(dict != NULL && name != NULL && answer != NULL)
    {
        item = getdns_dict_find(dict, name, false);
        if(item != NULL)
        {
            if(item->dtype != t_int)
                retval = GETDNS_RETURN_WRONG_TYPE_REQUESTED;
            else
            {
                *answer = item->data.n;
                retval = GETDNS_RETURN_GOOD;
            }
        }
    }

    return retval;
} /* getdns_dict_get_int */

/*---------------------------------------- getdns_dict_create */
struct getdns_dict *
getdns_dict_create()
{
    struct getdns_dict *dict;

    dict = (struct getdns_dict *) malloc(sizeof(struct getdns_dict));
    dict->rootp = NULL;

    return dict;
} /* getdns_dict_create */

/*---------------------------------------- getdns_dict_visit_copyitem */
/**
 * private function called by getdns_dict_copy() through the tree walk function 
 * is called as each node is visited.  twalk() calls 3x for each node and passes order
 * to tell us whether it is a pre/in/post order. We use this to copy the dictionary one item at
 * a time - this could be sped up
 */
void
getdns_dict_visit_copyitem(const void *node, VISIT order, int level)
{
    struct getdns_dict_item *item;

    item = *(struct getdns_dict_item **) node;
    /* postorder is mis-named - it results in in-order traversal */
    if(order == postorder || order == leaf)
    {
        switch(item->dtype)
        {
            case t_bindata:
                getdns_dict_set_bindata(walkresultdict, item->key, item->data.bindata);
                break;

            case t_dict:
                getdns_dict_set_dict(walkresultdict, item->key, item->data.dict);
                break;

            case t_int:
                getdns_dict_set_int(walkresultdict, item->key, item->data.n);
                break;

            case t_list:
                getdns_dict_set_list(walkresultdict, item->key, item->data.list);
                break;

            case t_invalid:
            default:
                // TODO: this is a fault of some kind, for now ignore it
                break;
        }
    }

    return;
} /* getdns_dict_visit_copyitem */

/*---------------------------------------- getdns_dict_copy */
/**
 * private function used to make a copy of a dict structure, the caller is responsible
 * for freeing storage allocated to returned value
 * NOTE: not thread safe - this needs to be fixed to be thread safe
 * @param srcdict the dictionary structure to copy
 * @return the address of the copy of the dictionary structure on success
 * @return NULL on error (out of memory, invalid srcdict)
 */
getdns_return_t
getdns_dict_copy(struct getdns_dict *srcdict, struct getdns_dict **dstdict)
{
    getdns_return_t retval = GETDNS_RETURN_NO_SUCH_DICT_NAME;

    if(srcdict != NULL && dstdict != NULL)
    {
        *dstdict = getdns_dict_create();
        walkresultdict = *dstdict;

        twalk(srcdict->rootp, getdns_dict_visit_copyitem);

        retval = GETDNS_RETURN_GOOD;
    }

    return retval;
} /* getdns_dict_copy */

/*---------------------------------------- getdns_dict_item_free */
/**
 * private function used to release storage associated with a dictionary item
 * @param all memory in this structure and its children will be freed
 * @return void
 */
void
getdns_dict_item_free(struct getdns_dict_item *item)
{
    if(item != NULL)
    {
        if(item->dtype == t_bindata)
        {
            if(item->data.bindata->size > 0)
                free(item->data.bindata->data);
            free(item->data.bindata);
        }
        else if(item->dtype == t_dict)
        {
            getdns_dict_destroy(item->data.dict);
        }
        else if(item->dtype == t_list)
        {
            getdns_list_destroy(item->data.list);
        }

        if(item->key != NULL)
            free(item->key);
        free(item);
    }
} /* getdns_dict_item_free */

/*---------------------------------------- getdns_dict_destroy */
void
getdns_dict_destroy(struct getdns_dict *dict)
{
    struct getdns_dict_item keyitem;
    struct getdns_dict_item *item;

    if(dict != NULL && dict->rootp != NULL)
    {
        while(dict->rootp != NULL)
        {
            item = *((struct getdns_dict_item **) dict->rootp);
            keyitem.key = item->key;
            tdelete(&keyitem, &(dict->rootp), getdns_dict_cmp);
            getdns_dict_item_free(item);
        }

        free(dict);
    }

    return;
} /* getdns_dict_destroy */

/*---------------------------------------- getdns_dict_set_dict */
getdns_return_t
getdns_dict_set_dict(struct getdns_dict *dict, char *name, struct getdns_dict *child_dict)
{ 
    struct getdns_dict_item *item;
    struct getdns_dict      *newdict;
    getdns_return_t retval = GETDNS_RETURN_NO_SUCH_DICT_NAME;

    if(dict != NULL && name != NULL)
    {
        item = getdns_dict_find(dict, name, true);
        if(item != NULL)
        {
            retval = getdns_dict_copy(child_dict, &newdict);
            if(retval == GETDNS_RETURN_GOOD)
            {
                item->dtype     = t_dict;
                item->data.dict = newdict;
            }
            else
                item->dtype = t_invalid;
        }
    }

    return retval;
} /* getdns_dict_set_dict */

/*---------------------------------------- getdns_dict_set_list */
getdns_return_t
getdns_dict_set_list(struct getdns_dict *dict, char *name, struct getdns_list *child_list)
{ 
    struct getdns_dict_item *item;
    struct getdns_list      *newlist;
    getdns_return_t retval = GETDNS_RETURN_NO_SUCH_DICT_NAME;

    if(dict != NULL && name != NULL)
    {
        item = getdns_dict_find(dict, name, true);
        if(item != NULL)
        {
            retval = getdns_list_copy(child_list, &newlist);
            if(retval == GETDNS_RETURN_GOOD)
            {
                item->dtype     = t_list;
                item->data.list = newlist;
            }
            else
                item->dtype = t_invalid;
        }
    }

    return retval;
} /* getdns_dict_set_list */

/*---------------------------------------- getdns_dict_set_bindata */
getdns_return_t
getdns_dict_set_bindata(struct getdns_dict *dict, char *name, struct getdns_bindata *child_bindata)
{ 
    struct getdns_dict_item *item;
    getdns_return_t retval = GETDNS_RETURN_NO_SUCH_DICT_NAME;

    if(dict != NULL && name != NULL && child_bindata != NULL)
    {
        item = getdns_dict_find(dict, name, true);
        if(item != NULL)
        {
            item->dtype  = t_bindata;
            item->data.bindata = (struct getdns_bindata *) malloc(sizeof(struct getdns_bindata));
            if(item->data.bindata != NULL)
            {
                item->data.bindata->data = (void *) malloc(item->data.bindata->size);
                if(item->data.bindata->data != NULL)
                {
                    item->data.bindata->size =  child_bindata->size;
                    memcpy(item->data.bindata->data, child_bindata->data, child_bindata->size);
                    retval = GETDNS_RETURN_GOOD;
                }
            }
        }
    }

    return retval;
} /* getdns_dict_set_bindata */

/*---------------------------------------- getdns_dict_set_int */
getdns_return_t
getdns_dict_set_int(struct getdns_dict *dict, char *name, uint32_t child_uint32)
{ 
    struct getdns_dict_item *item;
    getdns_return_t retval = GETDNS_RETURN_NO_SUCH_DICT_NAME;

    if(dict != NULL && name != NULL)
    {
        item = getdns_dict_find(dict, name, true);
        if(item != NULL)
        {
            item->dtype  = t_int;
            item->data.n = child_uint32;
            retval = GETDNS_RETURN_GOOD;
        }
    }

    return retval;
} /* getdns_dict_set_int */

/*---------------------------------------- getdns_dict_visit_print */
/**
 * private function called by the tree walk function invoked by getdns_pretty_print_dict
 * it is called as each node is visited.  twalk() calls 3x for each node and passes order
 * to tell us whether it is a pre/in/post order
 * TODO: need to handle nested non-trivial data types
 */
void
getdns_dict_visit_print(const void *node, VISIT order, int level)
{
    struct getdns_dict_item *item;
    int    newlen;
    char   *dtypestr = NULL;
    char   *valstr   = NULL;
    char   *itemstr  = NULL;

    item = *(struct getdns_dict_item **) node;
    /* postorder is mis-named - it results in in-order traversal */
    if(order == postorder || order == leaf)
    {
        switch(item->dtype)
        {
            case t_bindata:
                dtypestr = "bindata";
                valstr = strdup("NOT IMPLEMENTED");
                break;

            case t_dict:
                dtypestr = "dict";
                valstr = strdup("NOT IMPLEMENTED");
                break;

            case t_int:
                dtypestr = "int";
                asprintf(&valstr, "%d", item->data.n);
                break;

            case t_list:
                dtypestr = "list";
                valstr = strdup("NOT IMPLEMENTED");
                break;

            case t_invalid:
            default:
                dtypestr = "invalid";
                valstr = strdup("");
                break;
        }

        newlen = asprintf(&itemstr, "key=\"%s\", type=\"%s\", value=\"%s\"\n", item->key, dtypestr, valstr);
        if(newlen != -1)
        {
            walkresultchar = (char *) realloc(walkresultchar, walkresultcharlen + newlen + 1);
            memcpy(walkresultchar + walkresultcharlen, itemstr, newlen);
            walkresultcharlen += newlen;
            walkresultchar[walkresultcharlen] = '\0';
        }
        // else
            // TODO: this is a fault - do something

        free(valstr);
    }

    return;
} /* getdns_dict_visit_print */

/*---------------------------------------- getdns_pretty_print_dict */
char *
getdns_pretty_print_dict(struct getdns_dict *dict)
{
    char *retval = NULL;

    walkresultcharlen = 0;
    walkresultchar    = NULL;

    if(dict != NULL && dict->rootp != NULL)
    {
        twalk(dict->rootp, getdns_dict_visit_print);
        if(walkresultcharlen > 0)
        {
            retval = strdup(walkresultchar);
            free(walkresultchar);
        }
    }

    return retval;
} /* getdns_pretty_print_dict */

/* getdns_dict.c */
